Add cgroup_path built-in function

lenticularis39 opened the pull request:
The function takes a cgroup id and sends it to userspace via a perf
event, where it gets converted to a cgroup path.

<!--
Please provide a description of your change below this comment.

Then please complete the checklist.
-->

By combining the generic print mechanism with finding the cgroup path in userspace this provides an easy workaround to #1500, avoiding the possibility of an in-kernel deadlock. A limitation is that due to the conversion occuring in userspace, the cgroup path string cannot be manipulated.

I'm also not sure if `print_cgroup_path` is the best name for the function, maybe it would be better to name it `cgrouppath` to be consistent with other such printing functions like `join` that don't include "print" in their name.

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

fbs commented:
nice thanks for adding this!

I don't think we should add new printing functions to the language, they're annoying to use. E.g. what if I want to print the cgroup path along side with the pid? I'd have to do someting like

```
printf("Cgroup path: "); cgroup_path(..); printf(" PID: %d\n", pid)
```

Thats quite horrible and `cgroup_path` probably spits out a new line as well.

`time()` and `join()` also suffer from this.

I'd suggest something like we did to fix `time`, a new type that can only be used for printing. In the kernel it's just an integer you but when sent to userspace it gets translated properly. This is also how `strftime` works. The time formatting happens in userspace but you treat it just like a string in bpftrace code:

```
BEGIN { $ts = strftime("%m/%d/%y", nsecs); printf("%s\n", $ts); exit(); }
```

lenticularis39 commented:
I also thought about such solution, but I missed it already having been implemented for `time()`, I saw only the behavior of `join()` and thought that this problem should be fixed for all these functions in another PR. I'll look at how `strftime()` is implemented and try to use it to fix the cgroup path conversion.

lenticularis39 commented:
> I'll look at how `strftime()` is implemented and try to use it to fix the cgroup path conversion.

Done.

fbs commented:
looks good, will do a proper review later. Can you rebase on master to fix ci?

lenticularis39 commented:
Rebased.

viktormalik reviewed (commented):
Nice! LGTM, could you add some runtime test as well?

lenticularis39 commented:
Added runtime tests.

fbs commented on man/adoc/bpftrace.adoc:
```diff
@@ -1287,6 +1287,25 @@ t:syscalls:sys_enter_execve {
+
+Convert cgroup id to cgroup path.
+This is done asynchronously in userspace when the cgroup_path value is printed,
+therefore it can resolve to a different value if the cgroup id gets reassigned.
```
Just curious, how likely is this?

lenticularis39 replied:
Very unlikely, the reassignment would have to happen exactly between the two perf event handler calls, perhaps not possible at all if cgroup ids don't get reused.

viktormalik commented:
> It looks like the way I implemented the runtime tests depends on the cgroup path reported by /proc/.../cgroup to be an existing file in /sys/fs/cgroup, which is not the case in Alpine Linux (or maybe is /sys/fs/cgroup not mounted there?). 

Why not just run `bpftrace -e 'BEGIN { printf("%s", cgroup_path(cgroup)); }'` and check that it starts with `/sys/fs/cgroup`? I think that it could be sufficient here.

> Also the test will fail if run in an environment with no cgroup at all, since such cgroup id won't be found in /sys/fs/cgroup.

Similarly to the `cgroup` builtin [runtime test](https://github.com/iovisor/bpftrace/blob/master/tests/runtime/builtin#L137-L141), you can limit the test to the minimal kernel version that supports cgroups.

lenticularis39 commented:
> Why not just run `bpftrace -e 'BEGIN { printf("%s", cgroup_path(cgroup)); }'` and check that it starts with `/sys/fs/cgroup`? I think that it could be sufficient here.

The `cgroup` built-in returns an invalid value on my RHEL 8 VM for some reason:

```
$  sudo /opt/ktrace/bin/bpftrace -e 'BEGIN { print(cgroup); }'
Attaching 1 probe...
4294974185
^C
```

It might work on the CI, though, I'll try it.

lenticularis39 commented:
> Similarly to the `cgroup` builtin [runtime test](https://github.com/iovisor/bpftrace/blob/master/tests/runtime/builtin#L137-L141), you can limit the test to the minimal kernel version that supports cgroups.

That makes sense, although I assume it would still not work on systems that don't use cgroups (i.e. all their processes are in the root cgroup).

lenticularis39 commented:
> It might work on the CI, though, I'll try it.

It looks like it didn't work. (It does work on my openSUSE Leap 15.2 system, though.

lenticularis39 commented:
I'm reworking the code to better support cgroupv1 and mixed hierarchy (currently it just picks the cgroup path it finds first and prints it relative to /sys/fs/cgroup, which is likely not the expected behavior, e.g. compared to the contents of /proc/.../cgroup), so I'm marking it as draft for now. Will revert it when I'm done.

fbs commented:
woops forgot about this one. The code looks good on first glance :)

Is there a way I can help with tests?

lenticularis39 commented:
> woops forgot about this one. The code looks good on first glance :)

Thanks for looking at it. I updated it a bit, so now it should support all kinds of cgroup hierarchies (see the updated docs).

> Is there a way I can help with tests?

The issue appears to be related to cgroups on GitHub Actions, where bpftrace doesn't find the cgroup id in /sys/fs/cgroup, causing the runtime test for `cgroup_path` to fail. It doesn't happen when the tests are run locally in Docker.

A solution would be disabling the runtime test, but to do that properly it would be good to know the exact condition to check for (e.g. nothing is mounted at /sys/fs/cgroup) - I could try to run some checks inside the CI to see how it looks there. Do you have any other ideas what could be causing it?

fbs commented:
> 
> The issue appears to be related to cgroups on GitHub Actions, where bpftrace doesn't find the cgroup id in /sys/fs/cgroup, causing the runtime test for `cgroup_path` to fail. It doesn't happen when the tests are run locally in Docker.
> 
> A solution would be disabling the runtime test, but to do that properly it would be good to know the exact condition to check for (e.g. nothing is mounted at /sys/fs/cgroup) - I could try to run some checks inside the CI to see how it looks there. Do you have any other ideas what could be causing it?

Nope :(, also never looked at the insides of a actions VM.

lenticularis39 commented:
I figured it out, it's caused by Docker on cgroupv1 Ubuntu 20.04: the css_set's default cgroup is set to host containerd service, which is not present in the container's cgroup hierarchy:
```
root@69f5958e04f6:/home/tglozar/bpftrace# bpftrace -e 'BEGIN { printf("%d", cgroup); }'
Attaching 1 probe...
^C2071

tglozar@ubuntu-test:~$ find /sys/fs/cgroup/ -inum 2071
/sys/fs/cgroup/memory/system.slice/dbus.socket/memory.kmem.tcp.limit_in_bytes
/sys/fs/cgroup/unified/system.slice/containerd.service

root@69f5958e04f6:/home/tglozar/bpftrace# ls /sys/fs/cgroup/unified/system.slice/containerd.service /sys/fs/cgroup/memory/system.slice/dbus.socket/memory.kmem.tcp.limit_in_bytes
ls: cannot access '/sys/fs/cgroup/unified/system.slice/containerd.service': No such file or directory
ls: cannot access '/sys/fs/cgroup/memory/system.slice/dbus.socket/memory.kmem.tcp.limit_in_bytes': No such file or directory
```

This is a design flaw of the cgroupv1 hierarchy for Docker containers. I added a note about this into the manual page and added a condition to the runtime test that disables it when cgroupv2 filesystem is not mounted.

viktormalik commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -895,6 +895,28 @@ void CodegenLLVM::visit(Call &call)
+
+    // Store cgroup path event id
+    b_.CreateStore(b_.GetIntSameSize(cgroup_path_id_, elements.at(0)),
+                   b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(0) }));
```
Specify explicit type for `CreateGEP`, this version is deprecated on LLVM 13. See #2125.

viktormalik commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -895,6 +895,28 @@ void CodegenLLVM::visit(Call &call)
+    auto arg = call.vargs->at(0);
+    auto scoped_del = accept(arg);
+    b_.CreateStore(expr_,
+                   b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(1) }));
```
Same as above.

viktormalik commented on src/utils.cpp:
```diff
@@ -304,6 +304,121 @@ std::vector<std::string> get_kernel_cflags(
+    std::string filter)
+{
+  // TODO: Rewrite using std::views when C++20 support becomes common
+  auto roots = get_cgroup_hierarchy_roots();
```
What's the expected size of this vector? There's a lot of copying below so it could be more efficient to rewrite it by a loop.

lenticularis39 replied:
On my system there are 18 cgroup hierarchy roots, which is not much, nevertheless I could improve the code.

viktormalik commented on docs/reference_guide.md:
```diff
@@ -3092,6 +3093,25 @@ SRC 18:C0:4D:08:2E:BB, DST 74:83:C2:7F:8C:FF
+printing.
+
+Optionally a string literal may be passed as the second argument to filter cgroup hierarchies to look
+in (interpreted as a regex).
```
I'm wondering how complex we expect the regex expressions to be. If it's mostly about using `.*`, then it would be better to allow just `*` as a placeholder for a sequence of any chars (instead of full regexes) as we do in other places (e.g. probe matcher). You could use the `wildcard_match` function then.

lenticularis39 replied:
I don't expect them to be used too much, most of the time the user will probably want to use either `.*` or an exact name of the hierarchy. `wildcard_match` supports only `*`, though, and not something like `(unified|cpu)`, unlike the regex.

viktormalik replied:
I believe that we'll eventually get a support for that into our matcher so I'd still prefer going with `wildcard_match` for now. Users are also more used to specify path patterns using plain `*`, I'd say.

lenticularis39 replied:
OK, I'll move `get_tokens` to utils.cpp (currently it's private to probe_matcher.cpp) and swap the regex match with `wildcard_match` (which can be extended in another PR). I'd normally use glob(3) to match directly against the path, but cgroup2 can be mounted directly into `/sys/fs/cgroup`, which would break it...
